### PR TITLE
Implement Carbon-fiber Frames (Shelves) for BECal

### PR DESCRIPTION
**Pull request by Chao1009**
### Briefly, what does this PR introduce?
Separate BarrelCalorimeterInterlayers_geo into BarrelCalorimeterScFi_geo and BarrelCalorimeterImaging_geo.
Implement the carbon frame for Imaging (AstroPix) layers.
A few name changes to be consistent with ePIC BECAL naming scheme.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No, some names are changed (stave -> sector), but these changes were carefully selected so it does not break reconstruction.

### Does this PR change default behavior?
No, it does not change the default behavior.

### CODE REVIEW

File: src/BarrelCalorimeterSciFi_geo.cpp [hunk 1]
```diff
@@ -411,4 +411,4 @@ vector<FiberGrid> gridPoints(int div_n_phi, double div_dr, double trd_x1, double
   return results;
 }
 
-DECLARE_DETELEMENT(epic_EcalBarrelInterlayers, create_detector)
+DECLARE_DETELEMENT(epic_EcalBarrelSciFi, create_detector)
```
Comment (veprbl) [inline src/BarrelCalorimeterSciFi_geo.cpp:414]:
```suggestion
DECLARE_DETELEMENT(epic_EcalBarrelScFi, create_detector)
```
I thought, this was the current naming convention.

Comment (veprbl) [reply src/BarrelCalorimeterSciFi_geo.cpp:414]: (It's okay to change it, but we need to see where else to rename)

Comment (wdconinc) [reply src/BarrelCalorimeterSciFi_geo.cpp:414]: Not two days before a release please 🙏

Comment (veprbl) [reply src/BarrelCalorimeterSciFi_geo.cpp:414]: We can keep ScFi now, but decide on adopting SciFi for August.

**Comment by wdconinc** [comment]
@Chao1009 This is a draft and doesn't currently pass checks. Did you want to fix this and get this into the 23.07.0 tag?

**Comment by Chao1009** [comment]
> @Chao1009 This is a draft and doesn't currently pass checks. Did you want to fix this and get this into the 23.07.0 tag?

Yes, I'm currently working on it locally, should be ready soon.

**Comment by Chao1009** [comment]
With AstroPix layers (last column is thickness)
 |    72 CarbonFiber            6   11.968   1.5000     28.0746      51.2277      0.050
 |    73 Silicon               14   28.085   2.3300      9.3661      45.7531      0.050
 |    74 Silicon               14   28.085   2.3300      9.3661      45.7531      0.015
 |    75 Copper                29   63.546   8.9600      1.4356      15.6778      0.010
 |    76 Kapton                 6   12.701   1.4300     28.3749      55.9194      0.020
 |    77 Epoxy                  6   11.888   1.3000     31.5900      54.8420      0.010
 |    78 Air                    7   14.784   0.0012  30528.8407   71998.1725      1.795
 |    79 CarbonFiber            6   11.968   1.5000     28.0746      51.2277      0.050


Empty frame (last column is thickness)
 |    47 CarbonFiber            6   11.968   1.5000     28.0746      51.2277      0.050
 |    48 Air                    7   14.784   0.0012  30528.8407   71998.1725      1.900
 |    49 CarbonFiber            6   11.968   1.5000     28.0746      51.2277      0.050

**Comment by Chao1009** [comment]
@wdconinc ready for review

**Comment by wdconinc** [comment]
@Chao1009 Can you please remove all tabs? Several diffs are due to spaces converted to tabs.

File: src/BarrelCalorimeterImaging_geo.cpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+// 07/24/2021: Changed support implementation to avoid too many uses of boolean geometries. DAWN view seems to have
+//     issue dealing with it. C. Peng
+
+#include "DD4hep/DetFactoryHelper.h"
+#include "Math/Point2D.h"
+#include "TGeoPolygon.h"
+#include "XML/Layering.h"
+#include <functional>
+
+using namespace std;
```
Comment (wdconinc) [inline src/BarrelCalorimeterImaging_geo.cpp:23]: Avoid.

File: src/BarrelCalorimeterImaging_geo.cpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        // update values if there will be a frame
+        if (framing) {
+          auto x_frame = x_layer.child(_Unicode(frame));
+          l_thickness  = x_frame.attr<double>(_Unicode(height));
+          wt           = x_frame.attr<double>(_Unicode(thickness));
+          lenv_mat     = desc.material(x_frame.attr<std::string>(_Unicode(material)));
+          lfill_mat    = desc.material(x_frame.attr<std::string>(_Unicode(fill)));
+          // sanity check
+          if (l_thickness < (l_sthickness + 2*wt)) {
+            std::cerr << "EcalBarrelImaging_geo Error: frame available space " << l_thickness - 2.*wt
```
Comment (wdconinc) [inline src/BarrelCalorimeterImaging_geo.cpp:92]: Use Printout.

File: src/BarrelCalorimeterImaging_geo.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    double l_dim_y  = x_dim.z() / 2.;
+
+    // Loop over the sets of layer elements in the detector.
+    int l_num = 1;
+    for (xml_coll_t li(x_det, _U(layer)); li; ++li) {
+      xml_comp_t x_layer         = li;
+      int        repeat          = x_layer.repeat();
+      double     l_space_between = dd4hep::getAttrOrDefault(x_layer, _Unicode(space_between), 0.);
+      double     l_space_before  = dd4hep::getAttrOrDefault(x_layer, _Unicode(space_before), 0.);
+      bool       framing         = x_layer.hasChild(_Unicode(frame));
```
Comment (wdconinc) [inline src/BarrelCalorimeterImaging_geo.cpp:71]: Prefer booleans with prefix e.g. `has_frame` when used in isolation on `if` statements.

File: src/BarrelCalorimeterImaging_geo.cpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+          l_pos_z += l_space_between;
+        }
+        ++l_num;
+      }
+    }
+  }
+  // Phi start for a sector.
+  double phi = M_PI / nsides;
+  // Create nsides sectors.
+  for (int i = 0; i < nsides; i++, phi -= dphi) { // i is module number
```
Comment (wdconinc) [inline src/BarrelCalorimeterImaging_geo.cpp:171]: Why number sectors in negative phi direction? Seem like a recipe for confusion.

**Comment by Chao1009** [comment]
> 

Sure, I'll change tabs. I've changed from CentOS8 to Ubuntu22, but haven't dealt with vim yet so there are some tabs.

File: compact/ecal/barrel_interlayers.xml [hunk 6]
```diff
@@ -137,42 +137,34 @@
       </layer>
 
       <layer repeat="1" vis="EcalBarrelLayerVis"
-        space_before="EcalBarrel_ScFiLayerThickness + EcalBarrel_SpaceBetween">
-        <slice material="CarbonFiber" thickness="EcalBarrel_CarbonThickness" vis="EcalBarrelSliceVis"/>
-        <slice material="Air" thickness="EcalBarrel_ImagingLayerThickness_WithoutFrame " vis="EcalBarrelSliceVis"/>
-        <slice material="CarbonFiber" thickness="EcalBarrel_CarbonThickness" vis="EcalBarrelSliceVis"/>
+	space_before="EcalBarrel_ScFiLayerThickness + EcalBarrel_SpaceBetween">
```
Comment (wdconinc) [inline compact/ecal/barrel_interlayers.xml:140]: Remove tabs.

File: compact/ecal/barrel_interlayers.xml [hunk 7]
```diff
@@ -137,42 +137,34 @@
...
         <slice material="Epoxy" thickness="EcalBarrel_EpoxyThickness"         vis="EcalBarrelSliceVis"/>
-        <slice material="CarbonFiber" thickness="EcalBarrel_CarbonThickness" vis="EcalBarrelSliceVis"/>
       </layer>
 
       <layer repeat="1" vis="EcalBarrelLayerVis"
         space_before="EcalBarrel_ScFiLayerThickness + EcalBarrel_SpaceBetween">
-        <slice material="CarbonFiber" thickness="EcalBarrel_CarbonThickness" vis="EcalBarrelSliceVis"/>
-        <slice material="Air" thickness="EcalBarrel_ImagingLayerThickness_WithoutFrame " vis="EcalBarrelSliceVis"/>
-        <slice material="CarbonFiber" thickness="EcalBarrel_CarbonThickness" vis="EcalBarrelSliceVis"/>
+        <frame material="CarbonFiber" fill="Air" thickness="0.5*mm" height="2.*cm" vis="EcalBarrelSliceVis"/>
```
Comment (wdconinc) [inline compact/ecal/barrel_interlayers.xml:156]: Remove tabs.

File: compact/ecal/barrel_interlayers.xml (hunk 6 above)
Comment (wdconinc) [reply compact/ecal/barrel_interlayers.xml:140]:
```suggestion
        space_before="EcalBarrel_ScFiLayerThickness + EcalBarrel_SpaceBetween">
```

File: compact/ecal/barrel_interlayers.xml (hunk 7 above)
Comment (wdconinc) [reply compact/ecal/barrel_interlayers.xml:156]: Nevermind, this is fine.

**Review by wdconinc — APPROVED** [review]
For future reference, you can copy with history using this `git-cp <from> <to>` script: 
```
branch=$(mktemp --dry-run 'git-cp-XXXXXX')
git checkout -q -b $branch
git mv $1 $2
git commit -q -m "git-cp: move $1 to $2"
git checkout -q HEAD~ $1
git commit -q -m "git-cp: restore $1"
git checkout -q -
git merge -q -m "git-cp $1 $2" --no-ff $branch
git branch -D $branch
```

**Comment by wdconinc** [comment]
https://github.com/eic/epic/actions/runs/5470120067/jobs/9959869306?pr=470 failure is unrelated, and can be restarted when the rest of the (github) jobs have completed.

**Review by wdconinc — APPROVED** [review]

**Comment by wdconinc** [comment]
Passed overlap checks now.

**Pull request merged**